### ISSUE TITLE
ensure smooth apt experience around config processors with new m2e 2.x versions

### ISSUE THREAD

**Issue by martinlippert** [spring-projects/sts4#822]
The config processor support relies on the APT support to be configured correctly in the IDE. With the upcoming Eclipse 2022-09 release, which includes m2e 2.x, this support has changed and is disabled by default. We need to double check this latest version in combination with the config processor support to ensure a smooth user experience when shipping new Spring Tools based on Eclipse 2022-09 and beyond.

The latest snapshot builds on Eclipse 2022-09 from https://dist.springsource.com/snapshot/STS4/nightly-distributions.html already contain the latest m2e 2.x versions and have the new APT support for m2e 2.x installed by default.

**Comment by kdvolder** [comment]
Did some intial experimentation and I am hopeful this won't be a big deal to fix. 

If we manually enable the preference (shown in screenshot below) then things work much like they did before:

![image](https://user-images.githubusercontent.com/1002156/188999840-66442f8f-9bd0-4f18-acbf-54490281fbf7.png)

I suspect that the adoption of m2e-apt into Eclipse has simply changed the fq names of the configuration keys, which I would guess affects our code that attempts to turn these option on when we detect annotation processor being used in spring boot project.

If that's the case then it should be an easy thing to fix.

**Comment by kdvolder** [comment]
I did get an error popup switching the preference and also found this in Eclipse error log, which is not encouraging. Is this a problem with our packaging some old and new things together? Or is it a bug in m2e.apt / eclipse-jdt?

```
eclipse.buildId=4.16.0.202209071951
java.version=17.0.4.1
java.vendor=Eclipse Adoptium
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_CA
Framework arguments:  -product org.springframework.boot.ide.branding.sts4
Command-line arguments:  -os linux -ws gtk -arch x86_64 -product org.springframework.boot.ide.branding.sts4

org.eclipse.jface
Error
Wed Sep 07 16:27:54 PDT 2022
java.lang.NoSuchMethodError: 'org.eclipse.m2e.core.project.IMavenProjectFacade[] org.eclipse.m2e.core.project.IMavenProjectRegistry.getProjects()'

java.lang.NoSuchMethodError: 'org.eclipse.m2e.core.project.IMavenProjectFacade[] org.eclipse.m2e.core.project.IMavenProjectRegistry.getProjects()'
	at org.eclipse.m2e.apt.ui.preferences.AnnotationProcessingSettingsPage.getImpactedProjects(AnnotationProcessingSettingsPage.java:290)
	at org.eclipse.m2e.apt.ui.preferences.AnnotationProcessingSettingsPage.updateImpactedProjects(AnnotationProcessingSettingsPage.java:245)
	at org.eclipse.m2e.apt.ui.preferences.AnnotationProcessingSettingsPage.performOk(AnnotationProcessingSettingsPage.java:205)
	at org.eclipse.jface.preference.PreferenceDialog$7.run(PreferenceDialog.java:905)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.preference.PreferenceDialog.okPressed(PreferenceDialog.java:889)
	at org.eclipse.ui.internal.dialogs.FilteredPreferenceDialog.okPressed(FilteredPreferenceDialog.java:461)
	at org.eclipse.jface.preference.PreferenceDialog.buttonPressed(PreferenceDialog.java:233)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5796)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5025)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4477)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.internal.OpenPreferencesAction.run(OpenPreferencesAction.java:66)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5796)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5025)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4477)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
```

**Comment by kdvolder** [comment]
Possibly relevant commit that changed the 'missing' method from array return to return a List.

https://github.com/eclipse-m2e/m2e-core/commit/8b876a6748d8858578ee5894c40a872affc0e052

**Comment by kdvolder** [comment]
Relevant line of code in m2e-app-ui:
https://github.com/eclipse-m2e/m2e-core/blob/master/org.eclipse.m2e.apt.ui/src/org/eclipse/m2e/apt/ui/preferences/AnnotationProcessingSettingsPage.java#L290

This is where the error originates from (stacktrace). The source-code there however doesn't have any problem (it would work for either list or array... however the list vs array apis are not binary compatible, so this means of m2e-apt-ui was compiled against an older version of the api it would have runtime error.

The question is still whether this is an issue in our packaging, or whether its a problem with eclipse plugins themselves and how they were built.

**Comment by martinlippert** [comment]
Can you provide a few steps how to reproduce this from a fresh install of the latest nightly CI builds? That would be awesome!!!

**Comment by martinlippert** [comment]
From looking at the binary artifacts that we include, this looks like an issue with the official m2e release, it includes the exact same apt bundles that we have included in our builds. Let me see if I can find out more.

**Comment by martinlippert** [comment]
I filed this: https://github.com/eclipse-m2e/m2e-core/issues/896

**Comment by kdvolder** [comment]
I think you figure it out already but the steps are:
- start with a clean / new workspace
- open Preferences
- Go to Maven >> Annotation Processing
- Change annotation processing mode from 'Do not ...' to 'Automatically ...' 
- Click 'Apply And Close'
- Click 'Yes' on the popup that asks whether you want to rebuild projects after the change.

Observe:
- error popup 
![image](https://user-images.githubusercontent.com/1002156/189386139-166b8027-3470-49ed-b3c3-648381d79d6f.png)
- also the error is in the Eclipse error log.

**Comment by martinlippert** [comment]
The problem with the exception being thrown by m2e when changing the preferences around APT is fixed and released as part of m2e 2.0.5, which we include in the 4.16.0 release.

@kdvolder Do we need anything else from our side?

**Comment by kdvolder** [comment]
We probably still need something that turns on the setting from our side at the appropriate time (we have code somewhere that does this, but I reckon it doesn't work anymore because of some namespace changes in config property names).
